Support for module param defaults

agaffney opened the pull request:
##### SUMMARY
Support for module param defaults
Fixes #10849 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
playbook and modules

##### ANSIBLE VERSION
```
2.4
```


##### ADDITIONAL INFORMATION
This commit allows the definition of default params for a particular module. This is a very useful feature in Puppet that many people have asked for in Ansible.

Example:

```
---
- hosts: localhost
  gather_facts: no
  module_defaults:
    file:
      owner: foobar
  tasks:
    - file:
        path: /tmp/defaults_test
        state: touch
```

agaffney committed:
Support for module param defaults

agaffney commented:
@bcoca

agaffney commented:
Example of it working for an action plugin:

```
---
- hosts: localhost
  gather_facts: no
  vars:
    ansible_module_defaults:
      debug:
        msg: "testing defaults"
  tasks:
    - debug:
    - debug:
        msg: "overriding default"
    - debug:
        msg: "{{ omit }}"
```

Output:

```
TASK [debug] ******************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "testing defaults"
}

TASK [debug] ******************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "overriding default"
}

TASK [debug] ******************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "Hello world!"
}
```

agaffney commented:
Example with a module:

```
---
- hosts: localhost
  gather_facts: no
  vars:
    ansible_module_defaults:
      file:
        owner: foobar
  tasks:
    - file:
        path: /tmp/defaults_test
      ignore_errors: yes
    - file:
        path: /tmp/defaults_test
        owner: barbaz
      ignore_errors: yes
    - file:
        path: /tmp/defaults_test
        owner: "{{ omit }}"
```

Output:

```
TASK [file] *******************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "gid": 1000, "group": "agaffney", "mode": "0664", "msg": "chown failed: failed to look up user foobar", "owner": "agaffney", "path": "/tmp/defaults_test", "size": 0, "state": "file", "uid": 1000}
...ignoring

TASK [file] *******************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "gid": 1000, "group": "agaffney", "mode": "0664", "msg": "chown failed: failed to look up user barbaz", "owner": "agaffney", "path": "/tmp/defaults_test", "size": 0, "state": "file", "uid": 1000}
...ignoring

TASK [file] *******************************************************************************************************************************************************************
ok: [localhost]
```

agaffney commented:
Rebased for tests fix

agaffney commented:
Here's an example using the play attribute, rather than a variable, to set defaults:

```
---
- hosts: localhost
  gather_facts: no
  module_defaults:
    debug:
      msg: "default message"
  tasks:
    - debug:
    - debug:
        msg: override default message
    - debug:
        msg: "{{ omit }}"
```

Output:

```
TASK [debug] ********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "default message"
}

TASK [debug] ********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "override default message"
}

TASK [debug] ********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "Hello world!"
}
```

dagwieers commented:
You can already do something similar using YAML anchors and references.

agaffney commented:
Note: I don't think this will apply to the params for modules executed by action plugins, but that's probably a good thing.

agaffney commented:
Yes, I know, but you have to specify the anchor on each invocation of a module, where this is implicit.

dagwieers commented:
Well, I think there's something to say for making it explicit. Especially for supporting this.
So I would prefer not complicating Ansible with this functionality.

agaffney commented:
How exactly is this "complicating" Ansible? You're not forced to use it, and the additional code is simple an non-intrusive.

The YAML anchors trick only works within a single YAML file (afaik), and this specific feature has been requested many times in `#ansible`.

dagwieers commented:
Complicating in the sense that we have to support this functionality and any interactions it has.
Complicating support in the sense that if people copy&paste snippets we don't know that this mechanism is being used which makes it harder to provide support through these forums.

agaffney commented:
Potentially complicating support for users that don't provide enough context is not really a sufficient reason (imo) to block the addition of a useful and requested feature. It's no different in that regard than various other play-level attributes, such as `become` or `connection`.

dagwieers commented:
BTW One advantage of YAML anchors is that you can re-use them despite the module. So you can do something like this:

```yaml
---
- hosts: localhost
  gather_facts: no
  vars:
    stub: &owner
      owner: agafney
      group: agafney
  tasks:
    - file: &owner
        path: /foo/bar
        state: directory
    - copy: &owner
        src: foobar.txt
        dest: /foo/bar/
```

I am now abusing the vars-mechanism to create a dictionary to reuse as an example. Obviously this mechanism could be implemented some other way and a solution could be deviced to import definitions likewise.

dagwieers commented:
For `become` and `connection` there are no alternatives.

bcoca commented:
unlike a global config switch, this is less of an issue as the play/role will contain the defaults also

abadger commented on lib/ansible/playbook/play_context.py:
```diff
@@ -216,6 +217,9 @@ class PlayContext(Base):
     _fact_path        = FieldAttribute(isa='string', default=C.DEFAULT_FACT_PATH)
 
+    # Module defaults
+    _module_defaults  = FieldAttribute(isa='dict')
```
Use ```default={}``` here so that you don't have to compare against None in the rest of the code (unintuitive to a python programmer but I checked the FieldAttribute code and it does a deepcopy(default) in this case to make it work the right way).

abadger commented on lib/ansible/executor/task_executor.py:
```diff
@@ -488,6 +488,12 @@ def _execute(self, variables=None):
+        if self._play_context.module_defaults is not None and self._task.action in self._play_context.module_defaults:
+            tmp_args = self._play_context.module_defaults[self._task.action].copy()
+            tmp_args.update(self._task.args)
+            self._task.args = tmp_args
```
Does this not work?
``` python
self._task.args.update(self._play_context.module_defaults[self._task.action])
```

agaffney replied:
That merges the args in the wrong order. That applies the defaults on top of the user-specified params.

abadger commented on lib/ansible/playbook/play_context.py:
```diff
@@ -270,6 +274,9 @@ def set_play(self, play):
             self.force_handlers = play.force_handlers
 
+        if play.module_defaults is not None:
+            self.module_defaults = play.module_defaults
```
User experience question: Should this be merged instead of overridden?

alikins commented:
I would like to see a way to specify what name is used as the key in the module_defaults lookup. If not provided the module name (task.action) is used.

That would cover a lot of what the  'aliases/subclass' idea we discussed in IRC would. 

``` yaml
---
- hosts: localhost
  gather_facts: no
  module_defaults:
    debug:
      msg: "default message"
    banner_debug:
      msg: "================  default message =================="
    my_copy:
       mode: 0600
    copy_executable:
       mode: 0755
  tasks:
    # debug that uses the name of the action as the key in modules_default
    - debug:

    # debug that uses 'banner_debug' as the key in modules_default
    - debug:
        defaults_name: banner_debug

    - debug:
        msg: override default message
    - debug:
        msg: "{{ omit }}"

    # use copy, but with the 'my_copy' defaults
    - copy:
          src: /tmp/foo.txt
          dest: /tmp/new_foo.txt
      defaults_name: my_copy

    # use copy, without having the defaults changed at all
    - copy:
        src: /tmp/foo2.txt
        dest: /tmp/new_foo2.txt

    # use copy, but for executables 'preset'
    - copy:
        src: /tmp/myscript.sh
        dest: /usr/local/bin/myscript.sh
      defaults_name: copy_executable
```

This would allow:
1) A way to define a set of defaults without having to clobber all uses of the module (see 'copy' example above)
2) A way to have more than one set of default (or predefined) module params  (see 'copy_executable' and  'my_copy')

(no way attached to 'defaults_name' for the name here...)

alikins commented:
A variation that namespaces the defaults_name to per module name:

``` yaml
---
- hosts: localhost
  gather_facts: no
  vars:
    preset_mode: "script_install"
  module_defaults:
    debug:
      # if no 'default_name' is provided, the value for the 'default' key is used.
      default:
        msg: "default message"
      banner:
        msg: "================  default message =================="
   copy: 
     my_copy:
       mode: 0600
     executable:
       mode: 0755
  tasks:
    # debug that uses the name of the action as the key in modules_default
    - debug:

    # debug that uses 'banner' as the key in modules_default
    - debug:
        defaults_name: banner

    - debug:
        msg: override default message
    - debug:
        msg: "{{ omit }}"

    # use copy, but with the 'my_copy' defaults
    - copy:
          src: /tmp/foo.txt
          dest: /tmp/new_foo.txt
      defaults_name: my_copy

    # use copy, without having the defaults changed at all
    - copy:
        src: /tmp/foo2.txt
        dest: /tmp/new_foo2.txt

    # use copy, but for executables 'preset'
    - copy:
        src: /tmp/myscript.sh
        dest: /usr/local/bin/myscript.sh
      defaults_name: executable
```

Advantages over https://github.com/ansible/ansible/pull/22648#issuecomment-287119266

1) no name collisions for module_defaults keys (ie, if I accidentally used the name 'ping' for a default meant for copy, the flat approach would clobber the defaults for 'ping')

2) I could reuse default/preset names for the defaults of multiple modules. So in this version, I could create a presets called 'script_install' and 'doc_install' for each of 'copy', 'file', 'random_cloud_thing'. And
then choose the preset dynamically


``` yaml
---
- hosts: localhost
  gather_facts: no
  module_defaults:
    debug:
      # if no 'default_name' is provided, the value for the 'default' key is used.
      default:
        msg: "default message"
      banner:
        msg: "================  default message =================="
      script_install:
        msg: "------------------- script install ---------------------"
      key_install:
        msg: "XXXXXXXXX key install XXXXXXXXXXX"
   copy: 
     key_install:
       mode: 0600
     script_install:
       mode: 0755
  tasks:
    # debug that uses the name of the action as the key in modules_default
    - debug:

    # debug that uses '{{ preset_mode }}' as the key in modules_default
    - debug:
        defaults_name: "{{ preset_mode }}"

    - debug:
        msg: override default message
    - debug:
        msg: "{{ omit }}"

    # use copy, but with the preset_mode defaults
    - copy:
          src: /tmp/foo.txt
          dest: /tmp/new_foo.txt
      defaults_name: "{{ preset_mode }}"

    # use copy, without having the defaults changed at all
    - copy:
        src: /tmp/foo2.txt
        dest: /tmp/new_foo2.txt

    # use copy, but for executables 'preset'
    - copy:
        src: /tmp/myscript.sh
        dest: /usr/local/bin/myscript.sh
      defaults_name: "{{ preset_mode }}"
```


(This is starting to feel like an analog to inventory but for tasks....)

agaffney commented:
This new version seems to be slowing things down enough to cause test timeouts, probably due to the gymnastics that it goes through for each task to gather the defaults from the play, role, and any nested blocks.

abadger commented:
The only design issue I see is use of ```combine_vars()```.  Not sure if that should be used here or not.  Thoughts from @bcoca and @jimi-c?

abadger commented:
We think we want to use merge_hash() (not combine_vars) so that ansible_module_defaults from different soures are merged together rather than one overwriting another.

If we do that, we should document that whateverfield: "{{ omit }}" can be used to remove a default value

@bcoca should the module_defaults be settable in inventory as well?

bcoca commented:
no, its not connection related and that would impact 'what you execute' not 'how/where', the former we want to be obvious from looking at play.

agaffney commented:
@bcoca, @abadger

I've reworked this PR to greatly simplify it and added some tests around it. The `module_defaults` parameter now works much like the `environment` parameter (in the way that the values are merged between play, blocks, and task). It also now allows templating of the module defaults.

bcoca commented:
I do like this version a lot more

dagwieers commented:
I much rather have a solution that is not specific to parameters to one module. A lot of module-collections (think AWS, VMware, ACI, ...) could use a set of default values for the whole set of modules (connection parameters, tenant, etc.).

My proposal is here: https://github.com/ansible/proposals/issues/81
You basically have a collection of variables that modules subscribe to (so AWS related modules would subscribe to the aws collection). So it's similar than this, but with an abstraction layer rather than module-specific.

agaffney commented:
Frankly, that sounds overly complicated (both in code required for implementation and ongoing maintenance) for what can be easily accomplished with this PR and something like the following (using an example from your proposal):

```yaml
- hosts: localhost
  vars:
    aci_default_params:
      ...
  module_defaults:
    aci_tenant: '{{ aci_default_params }}'
    aci_vrf: '{{ aci_default_params }}'
```

The above is more explicit (and therefore easier to see at a glance what's going on) than just knowing that all AWS-related modules use the `aws` "bucket". Your proposal also seems to configure the feature via inventory vars, which has already been noted above in this PR as making it more difficult to see what's going on just from looking at the playbook.

dagwieers commented:
@agaffney Well, in most cases this information is coming from the inventory anyway and specific to groups or individual hosts. At least for the use-cases I describe (connection, tenant, default vmware hierarchy info, etc.) so that indirection is already there now (but repeated over the various tasks).

agaffney commented:
This PR allows defining the module defaults at the play, block, and task level. Since it also allows templating, you can easily define this mapping in a var in `group_vars/all.yml` (or similar) and then use `module_defaults: '{{ my_var }}'` in each play. This makes it reusable *and* explicit in the playbook.

jsumners commented:
I support this PR as a solution to my desire in https://github.com/ansible/ansible/pull/37478#issuecomment-375660033

dagwieers commented:
So the use-case that is not covered is the case where you have a bunch of modules needing the same parameters. Like AWS, or ACI. It's impractical to expect the user to repeat the same block for 50+ modules (and potentially missing new modules going forward) because you can only set parameters on a per-module basis. I think that property of this design is severely lacking.

Maybe registering for a specific variable namespace is not elegant, and that should be fixed. But this implementation is limiting in a way that you can't easily extend in the future. It will always be a per-module parameter set.

And the fact that this namespace can be set by the inventory is a very useful feature. You may call this "magic" or "hidden", but it's not different from setting your module parameters at the playbook/play level and then have roles acting out on it. If one is messy, the other is equally messy.

agaffney commented:
Despite how impractical or cumbersome it may be to use this PR for that particular use case, it's still possible in a straightforward (if somewhat annoying) way. Some of the pain can be mitigated by defining the dict to pass to `module_defaults` in `group_vars/all` and then doing `module_defaults: '{{ big_dict_with_module_defaults }}'`.

When I say "magic" or "hidden" in reference to your proposal, I mean from the perspective of trying to help users with issues in the `#ansible` IRC channel. The ability to have inventory vars (which typically aren't seen when a user shows their playbook) that have the ability to subtly modify the playbook's behavior in a way that can't be seen from looking at the playbook is bad for UX and bad for general supportability.

My design targets simplicity (at the expense of a little bit of verbosity), while yours targets "brevity" (for lack of a better word) at the expense of implementation/support complexity. Yours may be better for that particular use case, but mine is designed for the very general use case.

orthanc commented:
> And the fact that this namespace can be set by the inventory is a very useful feature. You may call this "magic" or "hidden", but it's not different from setting your module parameters at the playbook/play level and then have roles acting out on it. If one is messy, the other is equally messy.

Just one additional point in favour of this. Particularly in the context of AWS the modules already support "magic" mechanisms for providing credentials without specifying per-task, i.e. the boto environment variables and .aws/credentials file.

I don't see dagwieers proposal as any more magic than these existing mechanisms, but it does address the major issues that prevent me using these mechansims currently being:

1. I want to keep all the connection details in my inventory rather than have other depedendancies on the environment
2. I want to secure my credentials with ansible-vault which I can't do with the environment variables or .aws/credentials file
3. I want to be able to set these globally but still during my playbook (so that I can use temporary credentials after a 2FA authentication or assume_role)

dagwieers commented:
@agaffney Cumbersome, impractical, annoying all describe exactly how this use-case would not be realistic with your proposed mechanism. For that reason adding a new mechanism to Ansible that is limited to only specific modules would be a mistake. Harming future extensions by a short-sighted need and implementation.

agaffney commented:
@dagwieers

Is it even possible with your "bucket" proposal to apply defaults to a particular module rather than an entire "bucket" of modules? This PR is trying to solve a different problem than your proposal is, which is why it works in a different way.

I'm still waiting for you to actually give me a use case example that doesn't work with this PR. You've given one that is cumbersome (while trying to say that it doesn't work), but cumbersome and "is not supported" are very different things. I do not accept "I want to define this stuff outside of my playbook" as a use case, but more as a UX preference.

I'd appreciate it if you'd stop crapping all over my PR for hand-wavey "it doesn't do quite the same thing as *my* idea" reasons rather than actual technical reasons. You were crapping on it for different reasons (afaict, because you had submitted a PR that did this using vars that was ultimately rejected) when I created this a year ago, and I'm getting a bit tired of it.

dagwieers commented:
@agaffney Yes, look at alternatives at the very end, but I am open to an alternative that covers both use-cases or change my design. I am not married to my proposal. It's clear that you implementation solves a different problem, and my objection is that it's too limited. A better solution could cover more use-cases, but accepting this makes a better solution harder. So why would be limiting ourselves to a single module.

But let me be less diplomatic then: it is **impossible** to use your solution for the use-cases that a large set of users need. (At least 6 people added support for my use-case that doesn't fit your proposal). Even @alikins added it in this PR https://github.com/ansible/ansible/pull/22648#issuecomment-287119266 the same functionality we use buckets for. It looks a lot like my alternative proposal.

And I would appreciate if you didn't make up things. I never had a similar proposal that was rejected. There's no need for conspiracy theories or making things personal, even if you are getting a bit tired of the discussion. I can come up with theories why you are getting tired but that would not bring us any closer to something better that covers all use-cases.

dagwieers commented:
What is ironic in this whole discussion is the my initial opinion of this PR on 2017-03-05 was that this was not explicit enough and therefor 'hidden' magic. And [your point was that become and connection  has this property already](https://github.com/ansible/ansible/pull/22648#issuecomment-286847153).

> Potentially complicating support for users that don't provide enough context is not really a sufficient reason (imo) to block the addition of a useful and requested feature. It's no different in that regard than various other play-level attributes, such as become or connection.

Today our positions are reversed (you did convince me that this was already in Ansible in many places, even from the inventory, i.e. become and connection), and you now use my argument against my implementation, despite your implementation does the same thing which you already dismissed when it was my criticism.

I changed my opinion to support you in 2017, you changed your opinion to not support me in 2018.

dagwieers commented:
> Is it even possible with your "bucket" proposal to apply defaults to a particular module rather than an entire "bucket" of modules?

I never proposed my proposal for your use-case, but I am open to look at it. Your implementation is on the verge of being merged, which makes it more pressing to bring up objections to support other use-cases.

agaffney commented:
You keep saying that a certain use case is *impossible* but still won't say exactly what it is. The only one that I've heard about is not impossible with my PR, just somewhat cumbersome.

agaffney commented:
> And I would appreciate if you didn't make up things. I never had a similar proposal that was rejected.

Maybe I'm misremembering from a year ago, but I thought you had a similar PR that implemented pretty much this exact feature, but with variables instead of play/task attributes. I apologize if I was wrong or had the wrong impression.

abadger commented:
So reading this, I think that the additional use case is "A way to set defaults for groups of modules".  Is there anything other than that?

If so, I think there's two things we'd need:

1) A way to associate defaults to groups
2) A place to define which modules belong in which groups

For (1), could we take the UX in this PR, and add some marker for group names?  To steal from package managers:
```
module_defaults:
    ping:
        data: "Test"
    "@aws":
        username: "abc"
        password: "!vault123"
```
Note: "@" may not be the best thing to use as it means you have to quote the dict key.  We could use one special key in the toplevel like ```ansible_module_groups_``` instead, and then have the group_name: defaults mapping underneath that instead if that's preferred.

If this would work, then we *can* extend the framework in this PR to handle the group use case later.


For (2) I don't think that should be set in the modules.  (a) Users will have different groups in mind than the module author.. (I want to default mode for all modules that set mode on files.  I want to default state=latest for all packaging modules, I want to set validate_certs=False for all modules that have that, etc).  (b) Users may want to set certain defaults for all modules for a type but it may be that only a subset of the modules they care about have that parameter (hypothetically, imagine a set of cloud modules where all the modules take credentials as params but only a subset allow specifying a region.  If the user used some of the modules that did not specify a region, then they could not specify the region as a group default at all).

So group definition seems like something that we want to distribute a set of defaults but is easily extended by the user.  Perhaps a set of config files rather than inside of the modules?  Config format should have syntax for defining a new group, adding and removing a module to an existing group, and redefining an existing group from the ground up.

jsumners commented:
My use case is very simple: I want every instance of the `template` module in my playbook to have `lstrip_blocks: yes` (from PR #37478).  If I can further narrow it by `block` that'd be great, but I really want to be able to set it at the topmost level for the whole playbook.

agaffney commented:
The main part of this PR (in `lib/ansible/executor/task_executor.py`) is definitely reusable for any other framework that applies default parameters to modules. I have no strong feelings about if/how the "groups" stuff gets worked into this (in a separate PR) or the UX in general (except that it be *explicit* in a playbook or tasks file to define the defaults).

orthanc commented:
I was about to propose something similar. I do prefer the syntax in this PR as long as we can extend it to support the group case in future. I like abadger's proposal for how we might do that. Alternatively a `module_group_defaults` key as a sibling of `module_defaults` would achieve the same goal without need for a naming convention.

We'd need to define a precedence order between module_defaults and module_group_defaults but that's going to be needed regardless.

I actually do think that for many cases the module grouping belongs in the modules or at least bundled with ansible (AWS as a good example) to make the common cases easy and consistent. But I can see the merits of user defined groupings as well as a complementary capability. Of course that's a discussion for the subsequent extension not for this PR.

I do think the proposed mechanism in this PR will work well for defaults applied to single modules or small groups of modules (e.g. file & template or shell & command) so I do think adding it is an improvement and doesn't prevent adding a group mechanism in future.

dagwieers commented:
@abadger Thanks for making the opening. I do think we need to consider the UX (and capabilities) for grouping defined before merging this, just to ensure it all comes together well in playbooks. (We don't necessarily need to have it implemented though).

One important aspect is missing from the above, a way to set these from inventories. Since this is about connection information and other host or group-related information, I think it is relevant to be able to set this from the inventory as well without having to pass it per play (similar to become and connection are defined once).

agaffney commented:
The module parameters used to specify connection information to the network/clouid modules are not connection parameters in the traditional Ansible sense (in that they are used by ansible connection plugins or `become`). This PR (by itself, and with the module groups support) probably solves the problem for all cloud modules, and many of the network modules are already being moved to using connection plugins to solve it for that use case.

dagwieers commented:
@agaffney It would be nice if for once you are not minimizing my use-case. I haven't been doing that to yours either.

Fact is that there are much more modules that do not, and may never have connection plugins. And on top of that, the connection plugins do not factor in all connection-related parameters, at this point it *only* consists of ansible_user, ansible_password, ansible_host and ansible_port, but most HTTP/REST interfaces have plenty more that are custom and may never be supported (proxy, ssl, certificate, token, ...). And then we also have the non-connection related parameters (which overlaps with your use-case partly). But why am I defending my use-case again ?

That said, I am also part of making it easier for other modules to have HTTP-based connection plugins, so I care for both ends. https://github.com/ansible/proposals/issues/102#issuecomment-377630313

Neither is all-serving.

dagwieers commented:
Another feature I was considering for the "bucket" idea was to allow for more parameters to be set, that may not be supported by some of the modules. So some modules in a set don't need some of the information.

For example, ACI has modules that require a tenant, but other modules don't require a tenant) it would be nice if the mechanism would automatically set parameters that the module accepts, but skips others. So you can keep all those ACI parameters together without risking to break some modules.

I know VMware has the exact same thing, some modules require a datastore, VM folder, ESX host, cluster, but depending on what you're doing some of that information is not relevant. But every target node does have very specific information duplicated for these modules. If this would be automatically applied, this would be an even more powerful concept altogether.

agaffney commented:
I'm not trying to minimize your use case. It's more that I don't really understand it completely, since I'm not affected by it. From what I understand, that use case can be addressed by this PR (and the additional `module_group_defaults` support) by just requiring the addition of something like the following to the top of each play:

```
  module_defaults: '{{ defaults_from_var_in_inventory }}'
  module_group_defaults: '{{ group_defaults_from_var_in_inventory }}'
```

IMO, requiring this is much better than opening things up to the shenanigans and user confusion that would likely result from allowing setting these sort of parameters from outside the playbook. It just seems like much better UX, despite requiring those 1-2 additional lines in a play for your target use case.

bcoca commented:
@dagwieers connection plugins do support more than you state. For example winrm itself handles a slew of parameters more than the 'standard' ones you indicate.  There is nothing limiting the connection plugins from handling those parameter, they should just be connection related.

alikins commented:
A variation on this based on https://github.com/ansible/ansible/pull/22648#issuecomment-287130195

https://github.com/ansible/ansible/compare/devel...alikins:default_mod_args_22648_presets

an example playbook

``` yaml

---
- name: The first play
  hosts: localhost
  gather_facts: false
  # task_arg_presets?
  module_defaults:
    debug:
      default:
        verbosity: 1
      custom1:
        verbosity: 2
    file:
      default:
        owner: "adrian"
        mode: "u+x"
      other_preset:
        owner: "root"
        mode: "u=rwx,a=rx"
    copy:
      default:
        owner: "adrian"
        mode: "u+x"
      other_preset:
        owner: "root"
        mode: "u=rwx,a=rx"
    blorp:
      default:
        whatever: "This is arbitrary task name with a default set"
  vars:
      blip: foobar
  tasks:
    - name: blip
      tags:
        - the_task_tag
      debug:
        var: baz

    - name: test of file remove with some default module args
      file:
        path: /tmp/some_path
        state: directory

    - name: test of file remove with some default module args
      file:
        path: /tmp/some_path/status
        state: absent

    - name: test of copy with some default module for custom defaults_name other_preset
      copy:
        src: /proc/self/status
        dest: /tmp/some_path/status
      preset_name: "other_preset"

    - name: test of copy with some default module args
      copy:
        src: /proc/self/status
        dest: /tmp/some_path/status
```

agaffney commented:
@alikins I like the general idea of adding module "aliases" with default params set, but I'm not a fan of the UX from your branch.

Instead of requiring `preset_name` on the task, it may be simpler (from a user perspective) to just allow the alias to be the action in the task (`- my_debug:` for example). This wouldn't allow the use of duplicate aliases (such as you using `other_preset` for both `file` and `copy` above), but that's probably not a big deal. Requiring the `default` under the module name would also start to get a little weird when you're using `module_defaults` at the block/task level. That syntax makes sense at the play level, but it can start to get in the way elsewhere.

What about simplifying it a bit to separate the aliases and the defaults into separate attributes? This would allow keeping the current UX from this PR and keeping things a bit simpler at the block/task level.

```
- hosts: localhost
  module_aliases:
    - my_debug: debug
    - my_file: file
  module_defaults:
    debug:
      verbosity: 1
    my_debug:
      verbosity: 2
    my_file:
      mode: 0755
  tasks:
    - my_debug: msg="whatever"
    - my_file: path=/foo/bar state=touch
```

alikins commented:
@agaffney agreed on that particular yaml ds (in https://github.com/ansible/ansible/pull/22648#issuecomment-377985968) being weird.

bcoca commented:
discussed in irc meeting and majority agreed on merge.
